src/contracts/signing-key-events-cache/signing-key-events-cache.service.ts

```diff
@@ -0,0 +1,492 @@
+            {
+              count: moduleEvents.length,
+              address,
+            },
+          );
+        },
+      ),
+    );
+
+    return { events, startBlock, endBlock };
+  }
+
+  /**
+   * Retrieves signing key events data from the cache.
+   *
+   * This method fetches cached signing key events along with their associated headers.
+   * If the headers have default values (like 0 for the start and end block numbers),
+   * these values are updated to reflect the actual deployment block of the network.
+   *
+   * @returns {Promise<SigningKeyEventsCache>} A promise that resolves to a `SigningKeyEventsCache` object,
+   * containing the cached signing key events and their metadata.
+   */
+  public async getCachedEvents(): Promise<SigningKeyEventsCache> {
+    const { headers, data } = await this.levelDBCacheService.getEventsCache();
+
+    // default values is startBlock: 0, endBlock: 0
+    const deploymentBlock = await this.getDeploymentBlockByNetwork();
+
+    return {
+      headers: {
+        ...headers,
+        startBlock: Math.max(headers.startBlock, deploymentBlock),
+        endBlock: Math.max(headers.endBlock, deploymentBlock),
+      },
+      data,
+    };
+  }
+
+  /**
+   * Retrieves signing key events from the cache for the specified operators' keys.
+   *
+   * This method takes a list of operators' keys, ensures the list contains unique keys,
+   * and then fetches the corresponding events from the cache.
+   *
+   * @param {string[]} keys - An array of operators' keys for which to retrieve events.
+   * @returns {Promise<SigningKeyEventsCache>} Events associated with the specified keys.
+   */
+  public async getEventsForOperatorsKeys(
+    keys: string[],
+  ): Promise<SigningKeyEventsCache> {
+    const uniqueKeys = Array.from(new Set(keys));
+    return await this.levelDBCacheService.getCachedEvents(uniqueKeys);
+  }
+
+  /**
+   * Retrieves and returns all signing key events based on cached data and fresh data for a given key.
+   *
+   * This method combines cached signing key events with newly fetched events for a specific key,
+   * ensuring the cache is valid and updating the cache if necessary.
+   *
+   * @param {string} key - The specific signing key to retrieve events for.
+   * @param {number} blockNumber - The block number up to which the events should be retrieved.
+   * @param {string} blockHash - The block hash used to verify the integrity of the retrieved events.
+   * @returns {Promise<SigningKeyEventsGroupWithStakingModules>} merged signing key events and associated staking module addresses.
+   */
+  public async getUpdatedSigningKeyEvents(
+    key: string,
+    blockNumber: number,
+    blockHash: string,
+  ): Promise<SigningKeyEventsGroupWithStakingModules> {
+    const endBlock = blockNumber;
+    const cachedEvents = await this.getEventsForOperatorsKeys([key]);
+
+    const isCacheValid = this.validateCacheBlock(cachedEvents, blockNumber);
+    if (!isCacheValid) process.exit(1);
```

Comment:
⚠️ Shouldn't we just throw an exception in this case? So the duplicate checks do not affect the main functionality